tests/nn/data_parallel/test_fsdp_multiple_forward_checkpoint.py

```diff
@@ -310,39 +338,73 @@ def test_multiple_forward_checkpoint(precision, flatten, wrap_bn, model_type, bn
 
     world_size = 2
     expected_losses = None
-    # Ensure ddp == ddp+ckpt == fsdp == fsdp+ckpt.
+
+    # Ensure ddp == fsdp when modules are called multiple times per forward pass with/without checkpointing, forward
+    # counters and reducer bucketing.
+    #
+    # The bucketing check exists because the asynchronous gradient reduction it induces can interact with multiple
+    # forward passes in complex ways. For example, in the midst of a sharded backward pass, `parameter.grad` may only be
+    # `None` or an unsharded gradient tensor. The sharded tensor is then set at the end of the backwards pass. But a
+    # unit test with bucketing enabled might not catch violations of this invariant. For very small models, like the
+    # kind used in this unit test, bucketing will delay gradient reduction until after all the gradient computation is
+    # done. If the reduction incorrectly sets `.grad` to the _sharded_ variant, the test might not fail, since the
+    # gradient computations have already happened. Toggling bucketing helps verify that gradient reduction and
+    # computation interact correctly.
+    combinations = []
     for with_fsdp in [False, True]:
         for with_checkpoint in [False, True]:
             if not with_fsdp and with_checkpoint:
                 continue
-            final_losses = _get_cached_results(
-                world_size,
-                with_model2,
-                with_sync_bn,
-                with_fsdp,
-                with_checkpoint,
-                mixed_precision,
-                flatten,
-                wrap_bn,
-                fp32_reduce_scatter,
+            for with_fwd_counters in [True, False]:
```

Comment:
did this change make "fwd_counter" an optional thing now? If so, I think we should remove the fwd_counter from fsdp and checkpoint_wrapper since it only make things more complex. CC: @zhaojuanmao

Reply:
Yes, it's optional now. There may be a use case for it with checkpointed recurrent blocks, but I'll let others speak to that.

You may be able to achieve the same efficiency in other ways, like using no-sync contexts, or toggling reshard after forward. But then the post backward hook should capture more of the FSDP internal state, so it matches the forward pass. We do the same thing with the AMP state in activation checkpointing, for example.

Still setting that state correctly might be a bit confusing, since the first call would require sync, but not the ones after, due to the inverted order of execution during backward. The counters are a bit more intuitive, perhaps.

Reply:
interesting. what's the difference between recurrent blocks and block that have multiple fwd passes? I haven't worked on RNN's yet. Thanks again for this awesome work!

Reply:
Just the module call orders. RNNs would be like [A, A, A, ...], whereas my case is interleaved [A, B, C, A, B, C]. If the calls are consecutive, there's no point in resharding the full parameters. It wouldn't change the high water mark amount of vram used.

Reply:
Thanks. Do you mind just remove the fwd counter then? It would simplify the code and you don't need to test so many cases in the test as well.

Reply:
Done.